alt-tab-macos/ui/PreferencesPanel.swift

```diff
@@ -1,110 +1,211 @@
-            default:
-                throw "Tried to update an unknown popUpButton: '\(popUpButton)' = '\(popUpButton.titleOfSelectedItem!)'"
-            }
-            try! Preferences.saveRawToDisk()
+    private func makeLabelWithDropdown(_ keyPath: ReferenceWritableKeyPath<PreferencesPanel, NSPopUpButton?>, _ labelText: String, rawName: String, values: [String], suffixText: String? = nil) -> NSStackView {
+        let popUp = NSPopUpButton()
+        popUp.addItems(withTitles: values)
+        popUp.selectItem(withTitle: Preferences.rawValues[rawName]!)
+
+        self[keyPath: keyPath] = popUp
+
+        return makeLabelWithProvidedControl(labelText, rawName, popUp, suffixText)
+    }
+
+    private func makeLabelWithSlider(_ keyPath: ReferenceWritableKeyPath<PreferencesPanel, NSSlider?>, _ labelText: String, rawName: String, minValue: Double, maxValue: Double, numberOfTickMarks: Int, suffixText: String? = nil) -> NSStackView {
+        let slider = NSSlider(
+                value: Double(Preferences.rawValues[rawName]!)!,
+                minValue: minValue,
+                maxValue: maxValue,
+                target: self,
+                action: #selector(controlDidEndEditing)
+        )
+        slider.numberOfTickMarks = numberOfTickMarks
+        slider.allowsTickMarkValuesOnly = numberOfTickMarks > 1
+        slider.tickMarkPosition = .below
+
+        self[keyPath: keyPath] = slider
+
+        return makeLabelWithProvidedControl(labelText, rawName, slider, suffixText)
+    }
+
+    private func makeLabelWithProvidedControl(_ labelText: String, _ rawName: String, _ control: NSControl, _ suffixText: String? = nil) -> NSStackView {
+        let label = NSTextField(wrappingLabelWithString: labelText + ": ")
+        label.alignment = .right
+        label.widthAnchor.constraint(equalToConstant: calcLabelWidth()).isActive = true
+
+        control.identifier = NSUserInterfaceItemIdentifier(rawName)
+        control.target = self
+        control.action = #selector(controlDidEndEditing)
+        let containerView = NSStackView(views: [label, control])
+
+        if suffixText != nil {
+            let suffix = NSTextField(labelWithString: suffixText!)
+            suffix.textColor = .gray
+            containerView.addView(suffix, in: .leading)
         }
+
+        return containerView
     }
 
-    func textDidEndEditing(_ notification: Notification) {
-        if let textView = notification.object as? NSTextView {
-            let key = inputsMap[textView]!
+    /*
+    usage notes:
+    - NSSlider: supports on purpose currently only decimal values
+    */
+    @objc func controlDidEndEditing(senderControl: NSControl) {
+        self.makeFirstResponder(invisibleTextField) // deselects any possibly selected NSTextField (so slider & popUp changes deselect them)
+
+        let key: String? = senderControl.identifier?.rawValue
```

Comment:
You always set `identifier` on line 146 so you can use use `let key = senderControl.identifier!.rawValue` here

Reply:
Gotcha. I think I was writing defensive code there which is not needed nor useful in this scope.